### PR TITLE
Add QueueFount and unit tests

### PR DESCRIPTION
**Pull request by david415**
Dear Glyph,

Here's a QueueFount implementation I wrote this morning. What do you think?

### CODE REVIEW

**Comment by david415** [comment]
holy smokes i got it to pass all the tests!

File: tubes/queue.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+from zope.interface import implementer
+from twisted.internet import defer, task
+from twisted.python.failure import Failure
+
+from .kit import Pauser, beginFlowingTo
+from .itube import IFount, StopFlowCalled
+
+
+class NotABigTruckError(Exception):
```
Comment (david415) [inline tubes/queue.py]: ok... then what shall we call it?

File: tubes/queue.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    A series of tubes is not just a big truck you
+    can dump stuff onto...
+    """
+    pass
+
+
+
+@implementer(IFount)
+class QueueFount(object):
+    """
```
Comment (david415) [inline tubes/queue.py]: Should this description describe the implementation details of QueueFount?

File: tubes/queue.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
+# -*- test-case-name: tubes.test.test_queue -*-
+# Copyright (c) Twisted Matrix Laboratories.
+# See LICENSE for details.
+
+"""
+The QueueFount.
```
Comment (glyph) [inline tubes/queue.py]: This could use a better docstring :)

File: tubes/queue.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+from .kit import Pauser, beginFlowingTo
+from .itube import IFount, StopFlowCalled
+
+
+class NotABigTruckError(Exception):
+    """
+    A series of tubes is not just a big truck you
+    can dump stuff onto...
+    """
+    pass
```
Comment (glyph) [inline tubes/queue.py]: If you've got a docstring, you don't need a `pass`

File: tubes/queue.py (hunk 2 above)
Comment (glyph) [inline tubes/queue.py]: The whole point of the `series` API is that you can be totally ignorant of queuing theory and still write stuff that respects backpressure :)

File: tubes/queue.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+    def _actuallyResume(self):
+        """
+        Resume the flow (decrementing flowIsPaused).
+        """
+        self.flowIsPaused = False
+        self._turnDeque()
+
+
+    def push(self, item):
```
Comment (glyph) [inline tubes/queue.py]: This definitely should not be exposed as a method of `QueueFount`.  This is the single responsibility principle: each class should be responsible for just one thing. Which means it should implement just one public interface.  If we want an object with a `.push` method (and I'm not sure we do) we should have a public function that returns a 2-tuple of (thing-with-push) (`IFount`).

**Comment by glyph** [comment]
Thanks for your contribution!

I only have 5 minutes to comment right now, so this is just a quick question and then first impression:

What is the purpose of adding this class?

A great deal of the work within Tubes is designed to obviate this sort of API-that-might-raise-an-exception-or-might-not like `.push` or `.write` or `.append` .  This means it's the caller's responsibility to *not* call `.push` sometimes, and the system only *appears* to work if they've forgotten to do so.  The shape of `IFount` and `IDrain` is to force you to consider: where, externally, is your data coming from? How are you pulling it into the process?  and how do you *fully* apply backpressure to the *source* of the data, so that rather than "try to remember not to call `.push`" , you just call `.pauseFlow` when you can't handle more data and it propagates all the way to the edge of the system.

Hopefully this makes sense, but it's all just negative - if you can give me a better idea of your use-case, I can hopefully provide a bit of guidance as to the "tubesy" way to solve your problem.

**Comment by glyph** [comment]
Since, as it stands, I don't think this should be landed, I am going to close it.  However, that doesn't mean it's a permanent rejection - please feel free to re-open with an explanation of its purpose and the context that made you write it.

**Comment by david415** [comment]
I was reading your notes and they mention a QueueFount:
https://github.com/twisted/tubes/blob/master/sketches/notes.rst

I think lossy queues are useful. I was thinking to rewrite my onionvpn in tubes (right now it's written in twisted using producers and consumers); 
https://github.com/david415/onionvpn

Currently, onionvpn queues incoming packets from the tun device before sending out to the respective onion service connection.
https://github.com/david415/onionvpn/blob/master/onionvpn/ipv6_onion_consumer.py

**Comment by glyph** [comment]
@david415 Oh wow.  Sorry you implemented exactly what I asked for and I turned it down. Glad I wrote that down, so I can actually compare how my thinking has evolved :).

Upon consideration, some of what I wrote there I still think is valid: we definitely need a fount which consumes some in-memory value and then delivers it onward for testing.  We do need a way to push values through a test to ensure that they're handled at the appropriate time by the tube under test.  I do think that putting the "push" method on a separate object would be more in keeping with the SRP.

**Comment by glyph** [comment]
However, `tubes.memory` addresses a lot of those use-cases (written after that document, I'm sure), and I'm concerned that `QueueFount` will provide a temptation to do the wrong thing and just kinda hope the exception doesn't get raised rather than properly propagating backpressure.

**Comment by glyph** [comment]
The onion consumer is an interesting concrete use-case and I'll look at it.

**Pull request closed without merging**